### PR TITLE
Cluster

### PR DESCRIPTION
**Pull request by TJKoury**
First pull for rewrite.

### CODE REVIEW

File: cluster/index.js [hunk 1]
```diff
@@ -0,0 +1,183 @@
+#!/usr/bin/env node
+/**
+ * Copyright 2016 Lyteworx LLC.
```
Comment (chrisns) [inline cluster/index.js]: Pretty sure you'd need to set the copyright owner to `IBM Corp` not `Lyteworx LLC` if you wanted this to get merged as per the CLA.
Same applies for the headers for all your new files

File: cluster/index.js [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+            process.bingo = _stuckee.process.pid;
+            _stuckee.process.bingo = true;
+        }    
+    }
+    return process.bingo;      
+}
+
+if(cluster.isWorker){
+    process.consolebackup = console.log;
+    console.log = function(){}
```
Comment (chrisns) [inline cluster/index.js]: does this mean you'll be throwing away all the console output from workers?

File: cluster/index.js [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+            var _stuckee = clusterRED.getWorker();
+            process.bingo = _stuckee.process.pid;
+            _stuckee.process.bingo = true;
+        }    
+    }
+    return process.bingo;      
+}
+
+if(cluster.isWorker){
+    process.consolebackup = console.log;
```
Comment (chrisns) [inline cluster/index.js]: according to https://nodered.org/docs/writing-functions#logging-events
> node.log is the preferred method of logging

File: nodes/core/io/31-tcpin.js [hunk 4]
```diff
@@ -34,6 +34,7 @@ module.exports = function(RED) {
         this.server = (typeof n.server == 'boolean')?n.server:(n.server == "server");
         this.closing = false;
         this.connected = false;
+        this.cluster = native;
```
Comment (chrisns) [inline nodes/core/io/31-tcpin.js]: did you mean `"native"`?

File: nodes/core/io/32-udp.js [hunk 5]
```diff
@@ -28,6 +28,7 @@ module.exports = function(RED) {
         this.iface = n.iface || null;
         this.multicast = n.multicast;
         this.ipv = n.ipv || "udp4";
+        this.cluster = native;
```
Comment (chrisns) [inline nodes/core/io/32-udp.js]: did you mean `"native"`?

File: red/runtime/nodes/Node.js [hunk 6]
```diff
@@ -17,13 +17,17 @@
...
-
+var clusterRED = require("../../../cluster");
 var redUtil = require("../util");
 var Log = require("../log");
 var context = require("./context");
 var flows = require("./flows");
+var registry = require("./registry");
 
 function Node(n) {
+    //console.warn(n.type, 'node', process.pid, process.bingo);
```
Comment (chrisns) [inline red/runtime/nodes/Node.js]: don't commit commented code

File: red.js [hunk 7]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env node
 /**
- * Copyright 2013, 2016 IBM Corp.
+ * Copyright 2013, 2016 IBM Corp., 2016 Lyteworx LLC.
```
Comment (chrisns) [inline red.js]: as above CLA comment, a dual copyright will be even more complex to try and maintain

File: red/runtime/nodes/flows/index.js [hunk 8]
```diff
@@ -20,14 +20,15 @@ var when = require("when");
 var Flow = require('./Flow');
 
 var typeRegistry = require("../registry");
-var context = require("../context")
+var context = require("../context/")
```
Comment (chrisns) [inline red/runtime/nodes/flows/index.js]: probably don't need the added `/`?

File: settings.js [hunk 9]
```diff
@@ -69,7 +70,7 @@ module.exports = {
     // Node-RED scans the `nodes` directory in the install directory to find nodes.
     // The following property can be used to specify an additional directory to scan.
     //nodesDir: '/home/nol/.node-red/nodes',
-
+
```
Comment (chrisns) [inline settings.js]: whitespace?

File: settings.js [hunk 10]
```diff
@@ -102,14 +103,14 @@ module.exports = {
...
     // property can be used. See http://nodered.org/docs/security.html for details.
-    //adminAuth: {
-    //    type: "credentials",
-    //    users: [{
-    //        username: "admin",
-    //        password: "$2a$08$zZWtXTja0fB1pzD4sHCMyOCMYz2Z6dNbM6tl8sJogENOMcxWV9DN.",
-    //        permissions: "*"
-    //    }]
-    //},
+    adminAuth: {
```
Comment (chrisns) [inline settings.js]: does clustered node-red require authentication?

File: cluster/index.js (hunk 3 above)
Comment (TJKoury) [inline cluster/index.js]: Good catch!  Forgot to remove this.

File: cluster/index.js (hunk 1 above)
Comment (TJKoury) [inline cluster/index.js]: Right, will change this.  These are copies from our private repo, and our other fork will retain these copyright noticies.

File: cluster/index.js (hunk 2 above)
Comment (TJKoury) [inline cluster/index.js]: One of the discussion points is what to do with duplicate messages from each worker, like the startup console messages.

File: nodes/core/io/31-tcpin.js (hunk 4 above)
Comment (TJKoury) [inline nodes/core/io/31-tcpin.js]: Good catch!  I need to go through the other nodes and make sure I switched over per our new convention.

File: red/runtime/nodes/Node.js (hunk 6 above)
Comment (TJKoury) [inline red/runtime/nodes/Node.js]: Fixed!

File: red.js (hunk 7 above)
Comment (TJKoury) [inline red.js]: Will remove.

File: settings.js (hunk 10 above)
Comment (TJKoury) [inline settings.js]: It uses the same authentication as normal node-red.  I need to reset this settings file.

File: red/runtime/nodes/flows/index.js (hunk 8 above)
Comment (TJKoury) [inline red/runtime/nodes/flows/index.js]: Fixed!

File: cluster/index.js (hunk 2 above)
Comment (chrisns) [inline cluster/index.js]: IMHO pass it all up to the master process, output them with some sort of prefix

**Comment by TJKoury** [comment]
Closing this out due to lack of interest.

**Pull request closed without merging**